linkerd/app/outbound/src/http/logical/policy/route/backend.rs

```diff
@@ -131,6 +146,11 @@ where
                         })
                     }
                 }))
+                // Eagerly synthesize 504 responses for backend_request timeout
+                // errors.
+                // See the doc comment on `TimeoutRescue` for details.
+                .push(TimeoutRescue::layer())
+                .push_on_service(http::BoxResponse::<_>::layer())
```

Comment:
This is unrelated to this change but highlights something that may appear spooky to inner layers: the timeout layer will simply drop inner request futures, so inner layers (like the balancer) need to handle being dropped as a timeout of some sort. It would be better for us to move the timeout stack down into the endpoint (probably by fetching timeouts out of a request extension).

Reply:
Moving timeouts into the endpoint stack using a request extension was the initial design I used for https://github.com/linkerd/linkerd2-proxy/pull/2419 (see this commit https://github.com/linkerd/linkerd2-proxy/pull/2419/commits/64aca6a6fe6e562cc09a77443d2c7229f7c33a85). At the time, the conclusion in code review was that this approach was too complex. I'd be happy to bring that back, but I think it would make sense to do so in a separate PR?